### PR TITLE
WebRTC client & server

### PR DESCRIPTION
**Pull request by melekes**
Previous work by @tomaka: https://github.com/paritytech/smoldot/pull/1938
Refs #1712 

**TODO**:
- [x] read SDP spec https://datatracker.ietf.org/doc/html/rfc3264
- [x] modify SDP in webrtc server (and check one in client)
- [x] working connection
- [x] make necessary changes to `webrtc` rs library
- [x] sometimes DTLS handshake does not occur => debug & figure out why. maybe this is connected to `tls-id` attribute? no. it's because of the issue with local candidates (see https://github.com/melekes/libp2p-webrtc-direct/blob/ed829dc035b7cff1c155bfbd402261fbadf86d72/src/transport.rs#L531-L540)
- [x] implement https://docs.rs/libp2p/latest/libp2p/core/transport/trait.Transport.html
- [ ] submit a spec for WebRTC libp2p transport
- [ ] finish https://github.com/libp2p/rust-libp2p/pull/2622
- [ ] use the new transport in smoldot

```
ICE conn state: checking
main.js:1 ICE conn state: connected
main.js:1 conn state: connecting
main.js:1 'data' opened
main.js:1 conn state: connected
main.js:1 new message on 'data': 'WZaAkRgSkjiGwBw'
main.js:1 new message on 'data': 'dcgSNLysAvsyMJf'
main.js:1 new message on 'data': 'OZvQJpcHmpSdApM'
```

### CODE REVIEW

**Comment by melekes** [comment]
TODO:
- [ ] read SDP spec https://datatracker.ietf.org/doc/html/rfc3264
- [ ] modify SDP in webrtc server (and check one in client)
- [ ] working connection

Questions:
1. do we need multiple UDP ports for ICE traffic on the server?
2. is it possible to disable DTLS client's certificate verification? or all clients will have the same certificate too similar to servers?

**Comment by tomaka** [comment]
> do we need multiple UDP ports for ICE traffic on the server?

No

> is it possible to disable DTLS client's certificate verification? or all clients will have the same certificate too similar to servers?

I would prefer to disable DTLS client verification.
Hacks such as using the same certificate everywhere are necessary because we can't control the browser. However we're totally in control of the code on the server side, so we should do what is the most appropriate.

> modify SDP in webrtc server (and check one in client)

I'm not sure that's the right approach. If I'm not mistaken, the server doesn't need to manipulate SDP at all, unless it makes it easier to interact with the API of the `webrtc` library.

**Comment by melekes** [comment]
> I would prefer to disable DTLS client verification.

trying to figure out how to do it right now. It's possible in Go library, but not in webrtc-rs, which suggest we might need to tweak it.

> I'm not sure that's the right approach. If I'm not mistaken, the server doesn't need to manipulate SDP at all, unless it makes it easier to interact with the API of the webrtc library.

Yes for the most part. I was talking about modifying some fields like IP (of the offerer, ie. client) and media's timestamp

File: bin/wasm-node/javascript/src/webrtc/index.ts [hunk 1]
```diff
@@ -210,20 +210,30 @@ export default function(targetIp: string, protocol: 'tcp' | 'udp', targetPort: n
             // The maximum SCTP user message size (in bytes) (RFC8841)
             "a=max-message-size:100000" + "\n" +
             // A transport address for a candidate that can be used for connectivity checks (RFC8839).
-            "a=candidate:0 1 " + (protocol == 'tcp' ? "TCP" : "UDP") + " 2113667327 " + targetIp + " " + targetPort + " typ host" + "\n";
+            "a=candidate:1 1 " + (protocol == 'tcp' ? "TCP" : "UDP") + " 2113667327 " + targetIp + " " + targetPort + " typ host" + "\n" +
+            // Indicate that there will be no further candidates.
+            "a=end-of-candidates" + "\n";
```
Comment (tomaka) [inline bin/wasm-node/javascript/src/webrtc/index.ts:215]: From what I remember, I believe that technically you need to change the `a=ice-options:` attribute to the version that supports this

Comment (melekes) [reply bin/wasm-node/javascript/src/webrtc/index.ts]: right, it seems that it requires `ice-options:trickle`. removed

File: bin/webrtc-server/README.md [hunk 2]
```diff
@@ -0,0 +1,51 @@
+# WebRTC server
+
+## Install
+
+The certificate in `./static` directory **MUST** be marked as trusted (Chrome:
+'Security and privacy' -> 'Security' -> 'Manage certificates').
```
Comment (wemeetagain) [inline bin/webrtc-server/README.md:6]: Why is this necessary?

**Comment by melekes** [comment]
Closing. The work will be continued in https://github.com/libp2p/rust-libp2p/pull/2622

File: bin/webrtc-server/README.md (hunk 2 above)
Comment (melekes) [reply bin/webrtc-server/README.md:6]: You can discard that. It turned out to be another issue with the library I'm using.
